pkg/sharky/shard.go

```diff
@@ -5,187 +5,140 @@
+				case op = <-sh.reads:
+				default:
+					continue LOOP
 				}
-			} else {
-				// shard is allowed to grow upto limit
-				offset = size * DataSize
-				size++
 			}
-		}
-		// push the obtained offset to the main process
-		select {
-		case sh.free <- offset:
-		case <-sh.quit:
-			// remember free offset in limbo
-			sh.freed <- offset
-			return
-		}
-	}
-}
 
-// forever loop processing
-func (sh *shard) process() {
-	var writeOps chan *operation
-	var offset int64
-	free := sh.free
-	defer sh.wg.Done()
-	for {
-		select {
-		// prioritise read ops
-		case op := <-sh.readOps:
-			sh.read(op)
+			// only enabled if there is a free slot previously popped
+		case op := <-writes:
+			op.res <- sh.write(op.buf, slot)
+			free = sh.slots.out // reenable popping a free slot next time we can write
+			writes = nil        // disable popping a write operation until there is a free slot
+
+			// pop a free slot
+		case slot = <-free:
+			// only if there is one can we pop a chunk to write otherwise keep back pressure on writes
+			// effectively enforcing another shard to be chosen
+			writes = sh.writes // enable popping a write operation
+			free = nil         // disabling getting a new slot until a write is actually done
+
 		case <-sh.quit:
-			// this condition checks if an offset is in limbo (popped but not used for write op)
-			if writeOps != nil {
-				sh.freed <- offset
+			// this condition checks if an slot is in limbo (popped but not used for write op)
+			if writes != nil {
+				sh.slots.wg.Add(1) // Done after the slots process pops from slots.in
+				go func() {
+					sh.slots.in <- slot
+				}()
 			}
 			return
-		default:
-			select {
-
-			// pop a free offset
-			case offset = <-free:
-				// only if there is one can we pop a chunk to write otherwise keep back pressure on writes
-				// effectively enforcing another shard to be chosen
-				writeOps = sh.writeOps // enable popping a write operation
-				free = nil             // disabling getting a new position until a write is actually done
-
-				// only enabled if there is a free offset (pos) previously popped
-			case op := <-writeOps:
-				sh.write(op, offset)
-				free = sh.free // reenable popping a free slot next time we can write
-				writeOps = nil // disable popping a write operation until there is a free slot
-
-				// prioritise read operations
-			case op := <-sh.readOps:
-				sh.read(op)
-
-			case <-sh.quit:
-				// again put back offset in limbo
-				if writeOps != nil {
-					sh.freed <- offset
-				}
-				return
-			}
 		}
-
 	}
 }
 
+// close closes the shard:
+// wait for pending operations to finish then saves free slots and blobs on disk
 func (sh *shard) close() error {
-	go func() { // close free channel only after all relevant inflight change
-		sh.wg.Wait()
-		close(sh.freed)
-	}()
-	free := []int64{}
-	for offset := range sh.freed {
-		free = append(free, offset)
-	}
-	frees, err := json.Marshal(free)
-	if err != nil {
+	sh.slots.wg.Wait()
```

Comment:
i am starting to think that the complexity that is induced on `Close` in general probably should be mitigated on the localstore level, not on sharky. In the original forky implementation @janos used a `protect()` to protect the forky instance from shutdown, but I think that a similar thing can be applied on the localstore level in the following manner:
```
func Get() {
 db.protect()
 defer db.unprotect()
 select {
  case <-db.quit:
   return
 }
 ....
}
```
and vice versa on `Put`s. This way you can wait on the waitgroup in the localstore before calling the sharky close, and this way you get 100% assurance of a clean shutdown. However I wouldn't like to block on this now and we can refactor this later.